identity-account-storage/src/storage/tests.rs

```diff
@@ -0,0 +1,503 @@
+use identity_iota_core::document::IotaVerificationMethod;
+use identity_iota_core::tangle::MessageId;
+use identity_iota_core::tangle::Network;
+use identity_iota_core::tangle::NetworkName;
+
+use crate::identity::ChainState;
+use crate::types::KeyLocation;
+use crate::types::Signature;
+
+use super::Storage;
+
+macro_rules! ensure {
+  ($cond:expr, $($msg:expr),*) => {{
+    if !$cond {
+      let message: String = format!($( $msg, )*);
+      let fn_name: &'static str = function_name!();
+      return Err(anyhow::Error::msg(format!("[{}]: {}", fn_name, message)));
+    }
+  };};
+}
+
+macro_rules! ensure_eq {
+  ($left:expr, $right:expr, $($msg:expr),*) => {{
+    if $left != $right {
+      let message: String = format!($( $msg, )*);
+      let fn_name: &'static str = function_name!();
+      return Err(anyhow::Error::msg(format!("[{}]: {}", fn_name, message)));
+    }
+  };};
+}
+
+fn random_string() -> String {
+  rand::distributions::Alphanumeric.sample_string(&mut OsRng, 32)
+}
+
+#[named]
+pub async fn storage_did_create_test(storage: Box<dyn Storage>) -> anyhow::Result<()> {
+  let fragment: String = random_string();
+  let keypair: KeyPair = KeyPair::new(KeyType::Ed25519).unwrap();
+  let network: NetworkName = Network::Mainnet.name();
+
+  let expected_did: IotaDID = IotaDID::new_with_network(keypair.public().as_ref(), network.clone()).unwrap();
+  let expected_location: KeyLocation = KeyLocation::new(KeyType::Ed25519, fragment.clone(), keypair.public().as_ref());
+
+  let (did, location): (IotaDID, KeyLocation) = storage
+    .did_create(network.clone(), &fragment, Some(keypair.private().to_owned()))
+    .await
+    .context("did_create returned an error")?;
+
+  ensure_eq!(
+    did,
+    expected_did,
+    "expected returned did to be `{expected_did}`, was `{did}`"
+  );
+
+  ensure_eq!(
+    location,
+    expected_location,
+    "expected returned location to be `{expected_location}`, was `{location}`"
+  );
+
+  let exists: bool = storage
+    .key_exists(&did, &location)
+    .await
+    .context("key_exists returned an error")?;
+
+  ensure!(exists, "expected key at location `{location}` to exist");
+
+  let result: Result<_, crate::Error> = storage
+    .did_create(network, &fragment, Some(keypair.private().to_owned()))
+    .await;
+
+  ensure!(
+    result.is_err(),
+    "expected did_create to return an error when attempting to create an identity from the same private key twice"
+  );
+
+  let public_key: PublicKey = storage
+    .key_public(&did, &location)
+    .await
+    .context("key_public returned an error")?;
+
+  ensure_eq!(
+    public_key.as_ref(),
+    keypair.public().as_ref(),
+    "expected key_public to return `{:?}`, returned `{public_key:?}`",
+    keypair.public()
+  );
+
+  let network: NetworkName = Network::Devnet.name();
+  let (did, location): (IotaDID, KeyLocation) = storage
+    .did_create(network.clone(), &fragment, None)
+    .await
+    .context("did_create returned an error")?;
+
```

Comment:
One downside of having single test functions for e.g. `did_create` is that they each test multiple behaviours of those functions. Like here, if I got this error back as a `Storage` implementer I wouldn't know that this error was caused by the case where `None` is passed instead of a predefined `PrivateKey`.

Only options I can think of are: adding even more context to the errors, e.g. `.context("did_create with no private key failed")` or similar; or splitting the top-level test functions into e.g. `didCreateGeneratePrivateKeyTest` and `didCreateNoPrivateKeyTest`, or something.

Reply:
Split it in two, thanks for the suggestion!